File: VibroPhone/VibrationPlayer.m
//
//  VibrationPlayer.m
//  VibroPhone
//
//  Created by Juho Pispa on 12.3.2016.
//  Copyright © 2016 Juho Pispa. All rights reserved.
//

#import "VibrationPlayer.h"

void AudioServicesPlaySystemSoundWithVibration(int, id, id);

@implementation VibrationPlayer

- (void) playVibration:(NSArray *) times : (NSArray *) durations {
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
    NSMutableArray *arr = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < durations.count; i++){
        [arr addObject:[NSNumber numberWithBool:YES]];
        [arr addObject:[durations objectAtIndex:i]];
        if(i < (times.count)) {
            [arr addObject:[NSNumber numberWithBool:NO]];
            [arr addObject:[times objectAtIndex:i]];
        }
    }
    [dict setObject:arr forKey:@"VibePattern"];
    [dict setObject:[NSNumber numberWithInt:1] forKey:@"Intensity"];
    
    AudioServicesPlaySystemSoundWithVibration(4095,nil,dict);
}

@end

File: VibroPhone/VibrationPlayer.swift
//
//  VibrationPlayer.swift
//  VibroPhone
//
//  Created by Juho Pispa on 15.5.2016.
//  Copyright © 2016 Juho Pispa. All rights reserved.
//

import Foundation

var vibrationPlayer:VibrationPlayer = VibrationPlayer()

func playVibrationMessage(key: String, messages: [String:[String:Float]]) {
    let messageDict = messages[key]
    let keys = Array(messageDict!.keys)
    var messageDateDict = [NSDate:Float]()
    var times = [Int]()
    var durations = [Int]()
    
    for key in keys {
        let date = dateFromString(key)
        messageDateDict[date] = messageDict![key]
    }
    
    var dates = Array(messageDateDict.keys)
    dates.sortInPlace({ $0.compare($1) == NSComparisonResult.OrderedAscending })
    
    for date in dates {
        let duration = (messageDateDict[date]! * 1000)
        durations.append(Int(duration))
    }
    
    for var i = 1; i < dates.count; i++ {
        let time = (dates[i].timeIntervalSinceDate(dates[i-1]) * 1000)
        times.append(Int(time))
    }
    
    for var i = 1; i < times.count; i++ {
        times[i] = times[i] - durations[i-1]
    }
    
    vibrationPlayer.playVibration(times, durations)
}

func dateFromString(dateString: String) -> NSDate {
    let dateFormatter = NSDateFormatter()
    dateFormatter.dateFormat = Constants.dateFormat
    let date = dateFormatter.dateFromString(dateString)
    return date!
}